### PR TITLE
Add potential-unbounded-loop check patch and setup script

### PR DESCRIPTION
**Pull request by 1ntEgr8**
Summary:
This PR provides a setup script that patches LLVM (v11) with a `potential-unbounded-loop check`
- Diff that adds a `potential-unbounded-loop` check to `clang-tidy`
- Setup script that clones, builds, and patches LLVM

Test Plan:
- The diff includes two test files
    - `clang-tools-extra/test/clang-tidy/misc-potential-unbounded-loop.cpp`: Autogenerated using a python script (`clang-tools-extra/clang-tidy/misc/gen_potential_unbounded_loop_test.py`
    - `clang-tools/extra/test/clang-tidy/misc-potential-unbounded-loop-extra.cpp`: Tests for false positives

### CODE REVIEW

**Review by malfet — APPROVED** [review]

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 1]
```diff
@@ -0,0 +1,3051 @@
...
+diff --git a/clang-tools-extra/clang-tidy/rename_check.py b/clang-tools-extra/clang-tidy/rename_check.py
+index 4d5311c9a293..2410041fd5d2 100755
+--- a/clang-tools-extra/clang-tidy/rename_check.py
++++ b/clang-tools-extra/clang-tidy/rename_check.py
+@@ -120,16 +120,15 @@ def adapt_cmake(module_path, check_name_camel):
+       if (not file_added) and (cpp_line or cpp_found):
+         cpp_found = True
+         if (line.strip() > cpp_file) or (not cpp_line):
+-          f.write('  ' + cpp_file + '\n')
++          f.write(('  ' + cpp_file + '\n').encode())
```
Comment (malfet) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:366]: This does not seem related to the patch, does it?

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 2]
```diff
@@ -0,0 +1,3051 @@
...
++  modulecpp = next(filter(lambda p: p.lower() == module.lower() + 'tidymodule.cpp', os.listdir(module_path)))
+   filename = os.path.join(module_path, modulecpp)
+   with open(filename, 'r') as f:
+     lines = f.readlines()
+@@ -149,21 +148,21 @@ def adapt_module(module_path, module, check_name, check_name_camel):
+           header_found = True
+           if match.group(1) > check_name_camel:
+             header_added = True
+-            f.write('#include "' + check_name_camel + '.h"\n')
++            f.write(('#include "' + check_name_camel + '.h"\n').encode())
```
Comment (malfet) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:386]: Same, why those `.encode()` calls are needed? (And if they are, perhaps they could be added to a separate patch)

File: clang-tidy-checks/potential-unbounded-loop-check.diff (hunk 1 above)
Comment (1ntEgr8) [reply clang-tidy-checks/potential-unbounded-loop-check.diff:366]: I used this script to rename the check, but the script was buggy. Upstream LLVM had a fixed version, which I imported. I agree, it might be better to have a separate patch for this.

File: clang-tidy-checks/potential-unbounded-loop-check.diff (hunk 2 above)
Comment (1ntEgr8) [reply clang-tidy-checks/potential-unbounded-loop-check.diff:386]: I agree, it might be better to have a separate patch for this.

File: clang-tidy-checks/setup.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+
+function info() {
+  echo "[info]: " $1
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:4]: could you quote this? https://github.com/koalaman/shellcheck/wiki/SC2086
```suggestion
  echo "[info]: " "$1"
```

File: clang-tidy-checks/setup.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:1]: instead of using `&&` everywhere, could you `set -e` here?
```suggestion
#!/usr/bin/env bash
set -e
```

File: clang-tidy-checks/setup.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  echo "success!"
+}
+
+function check_requirements() {
+  info "checking requirements"
+  cmake --version &&
+  gcc --version &&
+  python3 --version &&
+  ninja --version &&
+  lld --version &&
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:17]: (see above comment about `set -e`)
```suggestion
  cmake --version
  gcc --version
  python3 --version
  ninja --version
  lld --version
```

File: clang-tidy-checks/setup.sh [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  success
+}
+
+function clone_llvm() {
+  info "cloing llvm"
+  git clone https://github.com/llvm/llvm-project.git &&
+  git fetch --all --tags &&
+  git checkout -b tags/llvmorg-11.0.0 &&
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:25]: (see above comment about `set -e`)
```suggestion
  git clone https://github.com/llvm/llvm-project.git
  git fetch --all --tags
  git checkout -b tags/llvmorg-11.0.0
```

File: clang-tidy-checks/setup.sh [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  success
+}
+
+function apply_patches() {
+  info "applying patches"
+  cd llvm-project
+  patch potential-unbounded-loop-check.diff &&
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:32]: (see above comment about `set -e`)
```suggestion
  patch potential-unbounded-loop-check.diff
```

File: clang-tidy-checks/setup.sh [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        -DLLVM_BUILD_TOOLS=OFF \
+        -DLLVM_BUILD_UTILS=OFF \
+        -DCMAKE_CXX_FLAGS_RELEASE="-O0" \
+        -GNinja ../llvm  &&
+  ninja
+}
+
+function setup() {
+  clone_llvm &&
+  apply_patches &&
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:56]: (see above comment about `set -e`)
```suggestion
  clone_llvm
  apply_patches
```

File: clang-tidy-checks/setup.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  build
+}
+
+check_requirements && setup
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:60]: (see above comment about `set -e`)
```suggestion
check_requirements
setup
```

File: clang-tidy-checks/setup.sh [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  cmake --version &&
+  gcc --version &&
+  python3 --version &&
+  ninja --version &&
+  lld --version &&
+  success
+}
+
+function clone_llvm() {
+  info "cloing llvm"
```
Comment (janeyx99) [inline clang-tidy-checks/setup.sh:22]:
```suggestion
  info "cloning llvm"
```

Comment (samestep) [inline clang-tidy-checks/setup.sh:22]: typo
```suggestion
  info "cloning llvm"
```

File: clang-tidy-checks/setup.sh [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  git clone https://github.com/llvm/llvm-project.git &&
+  git fetch --all --tags &&
```
Comment (samestep) [inline clang-tidy-checks/setup.sh:24]: you need a `cd` between these two, right? I think you might have to take the `cd llvm-project` from the `apply_patches` function and move it up here if I'm not mistaken

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 12]
```diff
@@ -0,0 +1,3051 @@
...
++  const auto *MatchedFor = Result.Nodes.getNodeAs<ForStmt>("for");
++
++  // only analyze non-compound types
++  if (LHSDecl->getType()->isCompoundType() ||
++      RHSDecl->getType()->isCompoundType() ||
++      IncDecl->getType()->isCompoundType()) {
++    return;
++  }
++
++  // only analyze loops in which increment var used in condition
```
Comment (janeyx99) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:127]:
```suggestion
+  // only analyze loops in which increment var is used in condition
```
nit

File: clang-tidy-checks/setup.sh (hunk 7 above)
Comment (samestep) [inline clang-tidy-checks/setup.sh:32]: also, once you `cd llvm-project`, wouldn't that mean the `potential-unbounded-loop-check.diff` file is no longer in the cwd, causing the `patch` to fail?

**Review by janeyx99 — APPROVED** [review]
some minor nits--great job!!

File: clang-tidy-checks/setup.sh (hunk 6 above)
Comment (samestep) [inline clang-tidy-checks/setup.sh:25]: I don't think this is correct (I think it just creates a branch from the current commit, and calls it `tags/llvmorg-11.0.0`); I believe what you want is to leave off the `-b`

File: clang-tidy-checks/setup.sh (hunk 7 above)
Comment (samestep) [inline clang-tidy-checks/setup.sh:32]: another question: I have this repo cloned in `~/github/pytorch/test-infra`, and I'm trying running this command from within my llvm/llvm-project clone (on commit `176249bd6732`):
```sh
patch ~/github/pytorch/test-infra/clang-tidy-checks/potential-unbounded-loop-check.diff
```
however, it's taking a long time and not give any output; if you write a `clang-tidy-checks/README.md`, could you indicate how long this command is expected to take?

**Review by samestep — COMMENTED** [review]
leaving some preliminary comments about `clang-tidy-checks/setup.sh`; I'm about to take a look at the main part of the diff right now :)

also, could you add a `clang-tidy-checks/README.md` that briefly says how to use these? e.g. what directory to be in when running `clang-tidy-checks/setup.sh`

File: clang-tidy-checks/setup.sh (hunk 7 above)
Comment (samestep) [reply clang-tidy-checks/setup.sh:32]: ok, I tried using `git apply` instead of `patch` and that seems to have worked; could you make that same replacement in `clang-tidy-checks/setup.sh`? also, it gave these warnings, in case that's relevant:
```
/Users/sestep/github/pytorch/test-infra/clang-tidy-checks/potential-unbounded-loop-check.diff:9: new blank line at EOF.
+
/Users/sestep/github/pytorch/test-infra/clang-tidy-checks/potential-unbounded-loop-check.diff:3051: new blank line at EOF.
+
warning: 2 lines add whitespace errors.
```

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 13]
```diff
@@ -0,0 +1,3051 @@
+diff --git a/clang-tools-extra/clang-tidy/cert/FloatLoopCounter.cpp b/clang-tools-extra/clang-tidy/cert/FloatLoopCounter.cpp
+index b6b3b1944d0e..ce966df15579 100644
+--- a/clang-tools-extra/clang-tidy/cert/FloatLoopCounter.cpp
++++ b/clang-tools-extra/clang-tidy/cert/FloatLoopCounter.cpp
+@@ -32,3 +32,4 @@ void FloatLoopCounter::check(const MatchFinder::MatchResult &Result) {
+ } // namespace cert
+ } // namespace tidy
+ } // namespace clang
++
```
Comment (samestep) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:9]: this seems irrelevant; if so, could you remove it?
```suggestion
```

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 14]
```diff
@@ -0,0 +1,3051 @@
...
++++ b/clang-tools-extra/docs/clang-tidy/checks/list.rst
+@@ -70,7 +70,7 @@ Clang-Tidy Checks
+    `bugprone-misplaced-widening-cast <bugprone-misplaced-widening-cast.html>`_,
+    `bugprone-move-forwarding-reference <bugprone-move-forwarding-reference.html>`_, "Yes"
+    `bugprone-multiple-statement-macro <bugprone-multiple-statement-macro.html>`_,
+-   `bugprone-no-escape <bugprone-no-escape.html>`_, "Yes"
++   `bugprone-no-escape <bugprone-no-escape.html>`_,
+    `bugprone-not-null-terminated-result <bugprone-not-null-terminated-result.html>`_, "Yes"
+    `bugprone-parent-virtual-call <bugprone-parent-virtual-call.html>`_, "Yes"
+    `bugprone-posix-return <bugprone-posix-return.html>`_, "Yes"
```
Comment (samestep) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:476]: what caused this change?

**Review by samestep — COMMENTED** [review]
main diff looks great, thank you! super impressive

File: clang-tidy-checks/setup.sh [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  gcc --version &&
+  python3 --version &&
+  ninja --version &&
+  lld --version &&
+  success
+}
+
+function clone_llvm() {
+  info "cloing llvm"
+  git clone https://github.com/llvm/llvm-project.git &&
```
Comment (driazati) [inline clang-tidy-checks/setup.sh:23]: You can do this all in one go with (also goes faster since it's only checking out the relevant branch)

```bash
git clone -b llvmorg-11.0.0 https://github.com/llvm/llvm-project.git --depth=1
```

File: clang-tidy-checks/setup.sh (hunk 7 above)
Comment (driazati) [inline clang-tidy-checks/setup.sh:32]: patch reads the patch file from stdin (probably why it's hanging) so you need to pipe it in:

```
cd llvm-project
cat ../potential-unbounded-loop-check.diff | patch -p1 -N -d .
```

File: clang-tidy-checks/setup.sh [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  patch potential-unbounded-loop-check.diff &&
+  success
+}
+
+function build() {
+  mkdir build
+  cd build
+  cmake -DCMAKE_C_COMPILER=clang \
+        -DCMAKE_CXX_COMPILER=clang++ \
+        -DCMAKE_BUILD_TYPE=Release \
```
Comment (driazati) [inline clang-tidy-checks/setup.sh:41]: RelWithDebInfo is probably better here since we don't really care about binary size and it's nice to have around in case of crashes

File: clang-tidy-checks/setup.sh [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        -DLLVM_ENABLE_PROJECTS=clang,clang-tools-extra \
+        -DLLVM_USE_LINKER=lld \
+        -DLLVM_TARGETS_TO_BUILD="X86" \
+        -DCLANG_ENABLE_STATIC_ANALYZER=OFF \
+        -DCLANG_ENABLE_ARCMT=OFF \
+        -DLLVM_BUILD_TOOLS=OFF \
+        -DLLVM_BUILD_UTILS=OFF \
+        -DCMAKE_CXX_FLAGS_RELEASE="-O0" \
```
Comment (driazati) [inline clang-tidy-checks/setup.sh:49]: This will speed up the build since it turns off optimizations but we probably don't want that for actual builds (since we want clang-tidy to go fast), I'd recommend leaving this off and for development you can set it as an environment variable instead

File: clang-tidy-checks/setup.sh [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        -GNinja ../llvm  &&
+  ninja
```
Comment (driazati) [inline clang-tidy-checks/setup.sh:51]: nit: Switch this to `cmake --build build`, it does the same thing but we don't have to change it for different build systems

**Review by driazati — COMMENTED** [review]
Awesome work! Here are some comments about the setup script, I'm getting to the main patch now

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 19]
```diff
@@ -0,0 +1,3051 @@
...
++    const MatchFinder::MatchResult &Result) {
++  const auto *LHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("lhs")->getDecl();
++  const auto *RHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("rhs")->getDecl();
++  const auto *IncDecl = Result.Nodes.getNodeAs<DeclRefExpr>("inc")->getDecl();
++  auto Op = Result.Nodes.getNodeAs<BinaryOperator>("op")->getOpcode();
++  auto IncOp = Result.Nodes.getNodeAs<UnaryOperator>("inc_op")->getOpcode();
++  const auto *MatchedFor = Result.Nodes.getNodeAs<ForStmt>("for");
++
++  // only analyze non-compound types
++  if (LHSDecl->getType()->isCompoundType() ||
```
Comment (driazati) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:121]: Should we give up for any value thats not an integer type (via `Type::isIntegerType()` instead? For something like this

```cpp
int main() {
    float max = 10000e30;
    for (unsigned long long i = 0; i < max; i++ ) {}
    return 0;
}
```

the widths of `max` and `i` are `32` and `64` respectively, but since it's a float with exponents the actual maximum value it takes on (that's used in the comparison) is much larger than `64` bits. This is a big rabbit hole we probably shouldn't be going down and should just give up early instead

File: clang-tidy-checks/potential-unbounded-loop-check.diff [hunk 20]
```diff
@@ -0,0 +1,3051 @@
...
++
++  auto LHSWidth =
++      LHSDecl->getASTContext().getTypeInfo(LHSDecl->getType()).Width;
++  auto RHSWidth =
++      RHSDecl->getASTContext().getTypeInfo(RHSDecl->getType()).Width;
++
++  if (LHSWidth < RHSWidth) {
++    if (((isLT(Op) || isLE(Op)) && isInc(IncOp)) ||
++        ((isGT(Op) || isGE(Op)) && isDec(IncOp))) {
++      diag(MatchedFor->getForLoc(), "Potential unbounded loop");
```
Comment (driazati) [inline clang-tidy-checks/potential-unbounded-loop-check.diff:141]: Could this be a little more verbose (preferably with some data)? Something like `Found LHS with size X and RHS with size Y, this can potentially lead to an unbound loop. Explicitly cast the RHS before comparing to fix this warning.`

**Review by driazati — APPROVED** [review]

**Review by driazati — CHANGES_REQUESTED** [review]
whoops I checked the wrong box, address the comments and this should be good to go

File: clang-tidy-checks/potential-unbounded-loop-check.diff (hunk 14 above)
Comment (1ntEgr8) [reply clang-tidy-checks/potential-unbounded-loop-check.diff:476]: It's the python script (that came with clang-tidy) to autogenerate boilerplate code. I didn't modify this file manually.

File: clang-tidy-checks/setup.sh (hunk 7 above)
Comment (1ntEgr8) [reply clang-tidy-checks/setup.sh:32]: Resolved below - I didn't pass in the directory in which the patch should be applied. This causes it to hang because it's waiting for input from stdin

**Review by driazati — APPROVED** [review]
🚀

File: clang-tidy-checks/README.md [hunk 21]
```diff
@@ -0,0 +1,11 @@
+# clang-tidy-checks
+
+This directory contains patches for additional `clang-tidy` that are not part of `llvm v11.0.0`. It includes a setup script that builds and applies the patches to `llvm v11.0.0`.
+
+## Usage
+
+```
+cd clang-tidy-checks
+chmod +x ./setup.sh
```
Comment (samestep) [inline clang-tidy-checks/README.md:9]: you actually don't need this, since the file permissions are tracked in Git:
```suggestion
```

**Review by samestep — APPROVED** [review]

File: clang-tidy-checks/README.md [hunk 22]
```diff
@@ -0,0 +1,11 @@
+# clang-tidy-checks
+
+This directory contains patches for additional `clang-tidy` that are not part of `llvm v11.0.0`. It includes a setup script that builds and applies the patches to `llvm v11.0.0`.
+
+## Usage
+
+```
```
Comment (driazati) [inline clang-tidy-checks/README.md:7]:
```suggestion
```bash
```

make sure to get the precious syntax highlighting

**Comment by zhouzhuojie** [comment]
@1ntEgr8 btw, I just added a docker base for clang-tidy, do you think the patch can be directly installed inside the dockerfile? https://github.com/pytorch/test-infra/blob/master/ci-docker-base/Dockerfile.cilint-clang-tidy

**Comment by 1ntEgr8** [comment]
@zhouzhuojie Yes, I can have the Dockerfile run the setup script. I'll submit a patch for it
